Disable StopWhenAllMixed feature

molnard opened the pull request:
Related to: https://github.com/zkSNACKs/WalletWasabi/issues/12602

This PR disables the `StopWhenAllMixed` feature. The change is not affecting the Manual CoinJoin mode. 

However in AutoCoin mode:

- Before this PR if the wallet is not entirely private (unconfirmed coins for example), we will keep going even if ALL the selected coins are private. 
- With this PR if the wallet is not entirely private (unconfirmed coins for example), we won't continue if ALL the selected coins are private, otherwise, we will continue. 

What happens after that? Are we restarting the client? The behavior was not changed by this PR - it is the same as before:

- In `AutoCoinJoin` mode we will keep trying no matter what, the wallet might receive new coins. 
- In ManualMode we will only stop if all coins are private, in any other case we will keep trying. 

Remarks

- See my remarks in the code as comments.

molnard commented on WalletWasabi/WabiSabi/Client/CoinJoin/Manager/CoinJoinManager.cs:
```diff
@@ -211,8 +211,8 @@ async Task<IEnumerable<SmartCoin>> SanityChecksAndGetCoinCandidatesFunc()
 					}
 
-					// If coin candidates are already private and the user doesn't override the StopWhenAllMixed, then don't mix.
-					if (coinCandidates.All(x => !x.Confirmed || x.IsPrivate(walletToStart.AnonScoreTarget)) && startCommand.StopWhenAllMixed)
```
`!x.Confirmed` can be ignored entirely. `SelectCandidateCoinsAsync` removes all the unconfirmed coins from the list. So it is not possible to have unconfirmed coins here. It will be false every time, so it does not matter in OR condition.

turbolay replied:
Where does `SelectCandidateCoinsAsync` remove all the unconfirmed coins from the list? I looked at the path and couldn't see it

molnard commented on WalletWasabi/WabiSabi/Client/CoinJoin/Manager/CoinJoinManager.cs:
```diff
@@ -211,8 +211,8 @@ async Task<IEnumerable<SmartCoin>> SanityChecksAndGetCoinCandidatesFunc()
 					}
 
-					// If coin candidates are already private and the user doesn't override the StopWhenAllMixed, then don't mix.
-					if (coinCandidates.All(x => !x.Confirmed || x.IsPrivate(walletToStart.AnonScoreTarget)) && startCommand.StopWhenAllMixed)
```
`startCommand.StopWhenAllMixed` is false the if condition cannot be satisfied - no matter what. It is false if `AutoCoinJoin` is turned **ON**. This is why the change will only affect AutoCoinJoin.

molnard commented:
I have tested it:

- Set the AutoCoinJoin to true.
- Set AnonScore target to have one unprivate coin. 
- Let the client run and wait until CJ.
- The unprivate coin gained score and became private.
- Client stopped registering coins. There were unconfirmed coins from the last CJ. 
- Changed the AnonScore target again to have at least one unprivate again.
- CoinJoin started again automatically.

turbolay commented:
There are other places where this feature is used, such as the `CoinJoinCoinSelector`.

Also, I thought that a good way to disable this feature without disabling it is to allow users to check a checkbox or something like it in the `Coinjoin Settings` to select `int.MaxValue` as `AnonScore Target`

molnard commented:
My purpose here was to fix the issue when the wallet kept mixing private coins infinitely - with a laser cut.

adamPetho reviewed (approved):
Such a small change, yet the consequence are not crystal clear for me.

I think this change is OK, so I'm approving this, but it would be nice if someone could test if it really fixes the issue.

@MarnixCroes could you perhaps run some rounds with this?

Also, in the future we really need to separate the concept "Auto Coinjoin" and "stop when all mixed", because right now they are handled with the same`CoinJoinTracker.StopWhenAllMixed` property and this is becoming the culprit of issues.

MarnixCroes reviewed (commented):
can confirm that wallet will not mix unnecessary when having an unconfirmed coin
(MusicBox says: "Insufficient funds eligible for coinjoin")

will test all scenario's later.

MarnixCroes commented:
it does unnecessary mixing when the non-private coin is below 5k sat
(it will coinjoin all the already private coins)

Szpoti commented:
> to select int.MaxValue as AnonScore Target
> Increasing the AnonScore target is a good next step

Wouldn't that mess up the privacy ring and percentage display pretty badly?
It would also be pretty bad UX for a user to mix and mix and mix for days & weeks, but only see 1% privacy, if it even reaches 1%.

molnard commented:
I was about to address @MarnixCroes comment and I noticed a misunderstanding that is causing misery around this.

I am not sure when it happened or how - but here is what I see. There has never been a `StopWhenAllMixed` feature. It is a variable used in the code. Have you ever seen a UI checkbox or something where you can turn on this feature? No because it is a logical indicator in the business logic. It indicates if AutoCoinJoin is ON or OFF. 

- If AutoCoinJoin is on (StopWhenAllMixed = false) the wallet will periodically check if there is something to mix. It makes sense, the wallet will never stop TRYing to mix. User might receives coins, etc. 
- The misunderstanding is `StopWhenAllMixed` means that we won't stop mixing coins ever. NO, false! 

So what now? I will start reverting all the changes related to this misunderstanding. 

Related: 

- https://github.com/zkSNACKs/WalletWasabi/pull/11762
- https://github.com/zkSNACKs/WalletWasabi/issues/12408
- https://github.com/zkSNACKs/WalletWasabi/pull/12502

molnard commented:
I have reverted: https://github.com/zkSNACKs/WalletWasabi/pull/11762 and added a test if small amounts are selected beside private coins - should not.

molnard commented on WalletWasabi/WabiSabi/Client/CoinJoin/Client/CoinJoinCoinSelector.cs:
```diff
@@ -80,7 +80,7 @@ public ImmutableList<TCoin> SelectCoinsForRound<TCoin>(IEnumerable<TCoin> coins,
 			.ToArray();
 
-		if (stopWhenAllMixed && semiPrivateCoins.Length + redCoins.Length == 0)
+		if (semiPrivateCoins.Length + redCoins.Length == 0)
```
This change is solving mixing below 5k stats. If there are no semiPrivateCoins or redCoins in the selection then CoinJoin will not continue.

molnard commented on WalletWasabi/WabiSabi/Client/CoinJoin/Client/CoinJoinCoinSelector.cs:
```diff
@@ -128,24 +128,6 @@ public ImmutableList<TCoin> SelectCoinsForRound<TCoin>(IEnumerable<TCoin> coins,
 		var orderedAllowedCoins = AnonScoreTxSourceBiasedShuffle(allowedCoins).ToArray();
 
-		// If the command is given to not stop when everything is coinjoined and the allowed private coins are empty, then we shortcircuit the selection.
-		if (!stopWhenAllMixed && allowedNonPrivateCoins.Count == 0)
```
This code was added for being able to select coins, even if all the coins in the wallet are private. We do not need that functionality, the reason mentioned above.

MarnixCroes reviewed (approved):
tack 8f0240143e1b5d4bfffbdedcacc5d41bac5577c3

part of what I tested:

- coinjoin doesn't start if the non-private coin is below 5k sat (https://github.com/zkSNACKs/WalletWasabi/issues/12536#issuecomment-2025001526)
- coinjoin doesn't start if the non-private coin is unconfirmed (https://github.com/zkSNACKs/WalletWasabi/issues/12536#issuecomment-1962363029)
- coinjoin doesn't start if the non-private coins are banned from CJ (https://github.com/zkSNACKs/WalletWasabi/issues/12536)
- tested normal behavior: auto-coinjoin on -> reach 100% -> coinjoin stops
- RPC: startcoinjoin with stopwhenallmixed=false -> coinjoin will not continue when all private (same as master)

anything else that should be tested?

molnard commented:
Thank you!
